### PR TITLE
1.7.7

### PR DESCRIPTION
**Pull request by AleshaOleg**

### CODE REVIEW

File: src/actions/channel.js [hunk 1]
```diff
@@ -161,13 +167,20 @@ export const openPm = (userId, options) => (dispatch, getState) => {
...
-    .then(([user, channel]) => {
-      dispatch(addUser(user))
-      dispatch(addChannel(channel))
+  let channelUsers
+
+  api
+    .openPm(org.id, userId)
+    .then(({id, users}) => {
+      channelUsers = users
+      return api.getChannel(id)
```
Comment (kof) [inline src/actions/channel.js]: use return Promise.all([users, api.getChannel(id)]).then(([users, channel]) => {})

File: src/actions/channel.js [hunk 2]
```diff
@@ -161,13 +167,20 @@ export const openPm = (userId, options) => (dispatch, getState) => {
...
-  ])
-    .then(([user, channel]) => {
-      dispatch(addUser(user))
-      dispatch(addChannel(channel))
+  let channelUsers
+
+  api
+    .openPm(org.id, userId)
+    .then(({id, users}) => {
+      channelUsers = users
```
Comment (kof) [inline src/actions/channel.js]: this is a really bad side effect

File: src/actions/common.js [hunk 3]
```diff
@@ -63,6 +65,37 @@ export const addUser = user => (dispatch) => {
...
 
+export const addNewUser = userId => (dispatch, getState) => {
+  const org = orgSelector(getState())
+
+  let channelUsers
+
+  return api
+    .openPm(org.id, userId)
+    .then(({id, users}) => {
+      channelUsers = users
```
Comment (kof) [inline src/actions/common.js]: same here

File: src/actions/pubSubHandlers.js [hunk 4]
```diff
@@ -173,22 +164,51 @@ export function handleLeftChannel({user: userId, channel: channelId}) {
...
+export function handleNotification(notification) {
+  return (dispatch, getState) => {
+    const state = getState()
+    const channels = channelsSelector(state)
+    const channel = find(channels, {id: notification.channelId})
+    if (channel) {
+      dispatch(newNotification(notification, channel))
+      return
+    }
+    dispatch(addNewUser(notification.author.id))
```
Comment (kof) [inline src/actions/pubSubHandlers.js]: dispatch should not return anything…

Comment (kof) [reply src/actions/pubSubHandlers.js]: I doubt that the return value from addNewUser action creator actually forwards the return to the dispatch function, but even if it does, to me dispatch() is like an even emitter, you should not expect any return values there.

File: src/actions/router.js [hunk 5]
```diff
@@ -61,7 +61,7 @@ export function goToChannel(channelOrChannelId, options) {
       // Assume we don't have always have all channels in the future.
       if (!channel) channel = {id: channelOrChannelId, slug: ''}
     }
-    const slug = channel.slug == null ? channel.mate.username : channel.slug
+    const slug = channel.slug == null ? channel.partner.username : channel.slug
```
Comment (kof) [inline src/actions/router.js]: hope you looked over the entire code based for "mate" and replaced it with "partner". Also such things are better done in a separate PR.

File: src/actions/sidebar.js [hunk 6]
```diff
@@ -30,6 +33,15 @@ export function showSidebar(show, options) {
   }
 }
 
+export const getUser = userId => (dispatch) => {
+  api
+    .getUser(conf.organization.id, userId)
+    .then((res) => {
+      dispatch(updateUserPartnerInfo(res))
+      dispatch(updateChannelPartnerInfo(res))
+    })
```
Comment (kof) [inline src/actions/sidebar.js]: unhandled error

File: i18n/translations/de.json [hunk 7]
```diff
@@ -78,8 +78,8 @@
   "_inviteMoreToGroup.comment": "Room Info Panel: link to invite people to the group/room",
   "addServiceIntegration": "Service Integration hinzufügen",
   "_addServiceIntegration.comment": "Room Info Panel: link to add an integration to the current room",
-  "welcomeToPm": "Private Konversation mit {mate}",
```
Comment (kof) [inline i18n/translations/de.json]: you actually can't update translations this way, you need to create a new key always.

Comment (kof) [reply i18n/translations/de.json]: what you did will cause you issues later, because translation tool will have old keys

File: src/actions/pubSubHandlers.js [hunk 8]
```diff
@@ -21,52 +23,64 @@ import {
...
+  const state = getState()
+  const channels = channelSelector(state)
+  const user = userSelector(state)
+
+  if (message.author.id === user.id || findIndex(channels, {id: message.author.id}) !== -1) {
+    dispatch(addNewMessage(message))
+    return
   }
+  dispatch(addNewUser(message.author.id))
+    .then(() => {
```
Comment (kof) [inline src/actions/pubSubHandlers.js]: same here

File: src/actions/pubSubHandlers.js [hunk 9]
```diff
@@ -173,22 +164,51 @@ export function handleLeftChannel({user: userId, channel: channelId}) {
...
-
-    dispatch({
-      type: types.REMOVE_USER_FROM_CHANNEL,
-      payload: {
-        channelId,
-        user,
-        isCurrentUser
-      }
-    })
+    const user = isCurrentUser ? currentUser : find(users, ({partner}) => partner.id === userId)
```
Comment (kof) [inline src/actions/pubSubHandlers.js]: there should not be such a thing as full list of users on the client.

File: src/actions/pubSubHandlers.js [hunk 10]
```diff
@@ -173,22 +164,51 @@ export function handleLeftChannel({user: userId, channel: channelId}) {
...
+      })
+    }
 
     const rooms = joinedRoomsSelector(getState())
     if (!rooms.length) dispatch(goTo('/chat'))
   }
 }
 
+const newNotification = (notification, channel) => (dispatch, getState) => {
+  const users = usersSelector(getState())
```
Comment (kof) [inline src/actions/pubSubHandlers.js]: same here, we don't have all users, it should be removed

File: src/components/sidebar/user-profile/UserProfile.js [hunk 11]
```diff
@@ -100,6 +105,11 @@ export default class UserProfile extends PureComponent {
     this.props.onShowSubview(tabs[index].name)
   }
 
+  userInfoClick = () => {
+    const {getUser, email, id} = this.props
+    if (!email) getUser(id)
```
Comment (kof) [inline src/components/sidebar/user-profile/UserProfile.js]: it should be always `onGetSomething`

File: src/actions/channel.js [hunk 12]
```diff
@@ -161,13 +167,17 @@ export const openPm = (userId, options) => (dispatch, getState) => {
...
-    api.getUser(org.id, userId),
-    api.openPm(org.id, userId)
-  ])
-    .then(([user, channel]) => {
-      dispatch(addUser(user))
-      dispatch(addChannel(channel))
+  api
+    .openPm(org.id, userId)
+    .then(({id, users}) =>
+      Promise.all([users, api.getChannel(id)])
```
Comment (kof) [inline src/actions/channel.js]: same line evtl?

File: src/actions/common.js [hunk 13]
```diff
@@ -63,6 +65,33 @@ export const addUser = user => (dispatch) => {
   })
 }
 
+export const addNewUser = userId => (dispatch, getState) => {
+  const org = orgSelector(getState())
+
+  return api
+    .openPm(org.id, userId)
+    .then(({id, users}) =>
+      Promise.all([users, api.getChannel(id)])
```
Comment (kof) [inline src/actions/common.js]: I would either put it onto same line or add braces ... need to start using prettier and verify this with eslint evtl

File: src/components/tags-input/TagsInput.js [hunk 14]
```diff
@@ -13,19 +14,30 @@ export default class TagsInput extends PureComponent {
...
+    focused: PropTypes.bool.isRequired,
+    list: PropTypes.array.isRequired,
+    value: PropTypes.string.isRequired,
+    placeholder: PropTypes.string.isRequired
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state = {filter: ''}
```
Comment (kof) [inline src/components/tags-input/TagsInput.js]: you can use initialiers without overwriting a constructor

File: src/components/tags-input/TagsInput.js [hunk 15]
```diff
@@ -52,55 +69,60 @@ export default class TagsInput extends PureComponent {
     this.props.onKeyDown(e)
   }
 
+  onChangeDebounce = debounce(() => {
```
Comment (kof) [inline src/components/tags-input/TagsInput.js]: onChangeDebounced

**Comment by nikgraf** [comment]
hey @kof, just looked into it with @AleshaOleg. Can you elaborate on or give an example on how you would do that?
> So far in order to reuse stuff I was extracting parts from dispatch into a separate reusable function.

I have clear idea how to deal with that in Redux saga, but since you explicitly need to wait for the `addNewUser` promise to finish this looks like a viable option. Of course it's really hard to test, but that's a general problem with thunks.

One option I was thinking of is constantly dispatching new message until the user is found. This works, but would pollute the Redux event queue. To avoid polluting the even queue with an invite loop it could have a delayed timing for the dispatch.

File: i18n/translations/de.json (hunk 7 above)
Comment (nikgraf) [reply i18n/translations/de.json]: @AleshaOleg is it correct that this is still an issue? Can you instead create a "welcomeToPrivateMessage" and a "privateMessageIntro" translation?

File: src/components/sidebar/user-profile/Contacts.js [hunk 16]
```diff
@@ -113,13 +113,14 @@ const Contacts = ({classes, email, skypeUsername, skypeForBusiness, phoneNumber}
 
 Contacts.propTypes = {
   classes: PropTypes.object.isRequired,
-  email: PropTypes.string.isRequired,
+  email: PropTypes.string,
```
Comment (nikgraf) [inline src/components/sidebar/user-profile/Contacts.js]: Why is email not required anymore?

File: src/actions/pubSubHandlers.js (hunk 9 above)
Comment (nikgraf) [reply src/actions/pubSubHandlers.js]: @AleshaOleg as far as I understand there is a list of users (not all of them) that is basically cached in the client. Was this the idea here? If so can you add a comment to the selector that this list necessarily doesn't include all the users of a particular organisation.

File: i18n/translations/de.json (hunk 7 above)
Comment (AleshaOleg) [reply i18n/translations/de.json]: @nikgraf, there is no issue. @kof thought I didn't make a pull request from lingohub. Everything ok

File: src/components/sidebar/user-profile/Contacts.js (hunk 16 above)
Comment (AleshaOleg) [reply src/components/sidebar/user-profile/Contacts.js]: Let's imagine, when we invited a new user to an organization and started a conversation with him. Then click to a sidebar, and want to see information about this user.

From out subscriptions events, we will receive info about the user - https://github.com/ubergrape/grape-web-client/blob/master/src/app/subscribe.js#L105 here.
But, without email. So, at first render, a component will not know about user email. That's why email not required anymore.

But, if you clicked on user profile page, you still have no email for this user. In this situation, we will get user info with the email from the server.
https://github.com/ubergrape/grape-web-client/blob/5bb5042df7fe7456a1ee728e768ae0182546f126/src/components/sidebar/user-profile/UserProfile.js#L110

Backend guys don't wanna to include email to a response from `organization.joined` because it's a pretty rare case where the user will click immediately to user profile page in the sidebar after the user was invited to the organization.

And in get_overview request email included. So, if you got a conversation with new user and refresh the page, you will receive this new user from get_overview.

Hope you understand my message:)

What you think about it?

File: src/actions/pubSubHandlers.js (hunk 9 above)
Comment (AleshaOleg) [reply src/actions/pubSubHandlers.js]: @nikgraf That's what I did here: https://github.com/ubergrape/grape-web-client/blob/release/src/selectors/index.js#L17

Maybe it's better to post this comment somewhere else?

File: src/components/sidebar/user-profile/Contacts.js (hunk 16 above)
Comment (nikgraf) [reply src/components/sidebar/user-profile/Contacts.js]: Makes sense! Thanks for the clarification 👍
